sui/src/rest_gateway.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::identifier::Identifier;
+use move_core_types::language_storage::TypeTag;
+use serde::de::DeserializeOwned;
+use serde_json::{json, Value};
+
+use sui_core::gateway_state::gateway_responses::TransactionResponse;
+use sui_core::gateway_state::GatewayAPI;
+use sui_types::base_types::{encode_bytes_hex, ObjectID, ObjectRef, SuiAddress};
+use sui_types::messages::{Transaction, TransactionData};
+use sui_types::object::ObjectRead;
+
+use crate::rest_gateway::requests::{
+    CallRequest, MergeCoinRequest, PublishRequest, SplitCoinRequest,
+};
+use crate::rest_gateway::responses::{NamedObjectRef, ObjectResponse, TransactionBytes};
+use async_trait::async_trait;
+pub mod requests;
+pub mod responses;
+
+pub struct RestGatewayClient {
+    pub url: String,
+}
+#[async_trait]
+#[allow(unused_variables)]
+impl GatewayAPI for RestGatewayClient {
+    async fn execute_transaction(
+        &mut self,
+        tx: Transaction,
+    ) -> Result<TransactionResponse, anyhow::Error> {
+        let url = format!("{}/api/execute_transaction", self.url);
+        let data = tx.data.to_base64();
+        let sig_and_pub_key = format!("{:?}", tx.tx_signature);
+        let split = sig_and_pub_key.split('@').collect::<Vec<_>>();
+        let signature = split[0];
+        let pub_key = split[1];
+
+        let body = json!({
+            "unsigned_tx_base64" : data,
+            "signature" : signature,
+            "pub_key" : pub_key
+        });
+        Ok(Self::post(url, body).await?)
+    }
+
+    async fn transfer_coin(
+        &mut self,
+        signer: SuiAddress,
+        object_id: ObjectID,
+        gas_payment: ObjectID,
+        recipient: SuiAddress,
+    ) -> Result<TransactionData, anyhow::Error> {
+        let url = format!("{}/api/new_transfer", self.url);
+
+        let object_id = object_id.to_hex();
+        let gas_payment = gas_payment.to_hex();
+        let value = json!({
+            "toAddress" : recipient.to_string(),
+            "fromAddress" : signer.to_string(),
+            "objectId" : object_id,
+            "gasObjectId" : gas_payment
+        });
+        let tx: TransactionBytes = Self::post(url, value).await?;
+        Ok(tx.to_data()?)
+    }
+
+    async fn sync_account_state(&self, account_addr: SuiAddress) -> Result<(), anyhow::Error> {
+        let url = format!("{}/api/sync_account_state", self.url);
+        let client = reqwest::Client::new();
+        let address = account_addr.to_string();
+        let body = json!({ "address": address });
+
+        client.post(url).body(body.to_string()).send().await?;
+        Ok(())
+    }
+
+    async fn move_call(
+        &mut self,
+        signer: SuiAddress,
+        package_object_ref: ObjectRef,
```

Comment:
nit: since only the ObjectID is required
```suggestion
        package_object_ref: ObjectID,
```

Reply:
We have plan to refactor the GatewayAPI and use ObjectRef in all endpoints, so the user will know exactly which version of the object they will be using, currently I am using object id only because it is easier to serialise, will fix this in next PR.